### PR TITLE
nrfx: Use actual nrfx instead of manually modified files

### PR DESCRIPTION
**Pull request by rugeGerritsen**
With nrfx 2.3.0 it is possible to include the MDK files on
non-ARM platforms. As the MDK has a depenency to cmsis core
files, a simple mocked core_cm4.h has been provided.

By defining NRF_DECLARE_ONLY and NRF_STATIC_INLINE to nothing,
no functions are inlined. This ensures that it is possible to
mock the peripherals.

For a non-zephyr build, the application must provide the
environment varianble NRFX_BASE.

The user of the nRF MDK or nrfx must ensure that the peripheral
pointers are redefined before they are used. This is done in
nrfx_bsim_redef.h. Include this header after including nrf.h
or similar. When using nrfx, this inclusion can be placed in
nrfx_glue.h

For non-zephyr builds, a nrfx_config and nrfx_glue is provided.

The original implementation of the HAL APIs from the real nrfx
is included in `nrf_hal_originals.c`.
All these functions are labelled as weak,
so they can be replaced as needed with new definition in `src/nrfx/hal/`
This removes the need for duplicating code.

### CODE REVIEW

**Comment by rugeGerritsen** [comment]
@anangl , does this fit your usecase?

**Comment by anangl** [comment]
> @anangl , does this fit your usecase?

You mean the added `nrf_power.c`? Yes. Actually, after https://github.com/zephyrproject-rtos/zephyr/pull/27980 gets merged, no function from `nrf_power.h` will be needed (currently the `nrfx_power.c` driver is unnecessarily compiled for the nrf52_bsim target, as the nRF clock control driver in Zephyr in fact does not use it).
But if you aim to cover all functions that could be called from `nrfx_power.c`, you'll need to add also implementations of `nrf_power_int_enable()` and `nrf_power_int_disable()`.

**Comment by rugeGerritsen** [comment]
> > @anangl , does this fit your usecase?
> 
> You mean the added `nrf_power.c`? Yes. Actually, after [zephyrproject-rtos/zephyr#27980](https://github.com/zephyrproject-rtos/zephyr/pull/27980) gets merged, no function from `nrf_power.h` will be needed (currently the `nrfx_power.c` driver is unnecessarily compiled for the nrf52_bsim target, as the nRF clock control driver in Zephyr in fact does not use it).
> But if you aim to cover all functions that could be called from `nrfx_power.c`, you'll need to add also implementations of `nrf_power_int_enable()` and `nrf_power_int_disable()`.

Initially I just wanted to add the minimal functionality to be able to use bsim with nrfx.

File: src/HW_models/irq_ctrl.c
Comment (aescolar) [inline src/HW_models/irq_ctrl.c]: :) , I guess we do not need to change this file just for this

**Comment by aescolar** [comment]
@rugeGerritsen (I was on holidays these last 2 weeks, so I'm quite out of context) Is this PR still relevant?

**Comment by rugeGerritsen** [comment]
Hi @aescolar , yes, this PR is still relevant. This PR will simplify future maintenance. 
This PR verifies the integration: https://github.com/zephyrproject-rtos/zephyr/pull/27882.

File: Makefile
Comment (aescolar) [inline Makefile:23]: To be nice to users, please add something like:

```diff
diff --git a/Makefile b/Makefile
index 983f128..3c521e0 100644
--- a/Makefile
+++ b/Makefile
@@ -8,6 +8,10 @@ include ${BSIM_BASE_PATH}/common/pre.make.inc
 
 SRCS:=$(shell find src/ -name "*.c")
 
+ifndef NRFX_BASE
+$(error NRFX_BASE needs to be set to point to bablablabl)
+endif
+
 INCLUDES:=-I${libUtilv1_COMP_PATH}/src/ \
           -I${libPhyComv1_COMP_PATH}/src/ \
           -I${2G4_libPhyComv1_COMP_PATH}/src \
```

Comment (aescolar) [reply Makefile:23]: Also please update docs/README.md , so it explains that it uses the real nordic HAL. That to build it without Zephyr, the nRF HAL must be fetched in some place, and the environment variable NRFX_BASE must be set to point to it.

File: src/nrfx/hal/nrf_power.c
Comment (aescolar) [inline src/nrfx/hal/nrf_power.c:11]: These 4 functions, can they not be directly the ones from the real HAL?

File: CMakeLists.txt
Comment (aescolar) [inline CMakeLists.txt:25]: How do you ensure the headers provided here are included instead of the ones from the nRFx west module? (both provide nrfx_config.h and nrfx_glue.h)

**Review by aescolar — DISMISSED** [review]
You need to sign off all commits :-p

If I understand this PR, the target would be to use the HAL headers from the real nRFx, but only for the prototypes, and then have all interface bodies replaced in the .c files. Correct?
(That results in a lot of copy-pasting still. If we are able to start including the real HAL files, it would be much nicer if we could avoid it)
Have you considered having all the original headers functions be compiled in but as weak (`__attribute__((weak))`), and then only providing a replacement implementation in the c files for the ones that need replacing?

File: Makefile
Comment (rugeGerritsen) [reply Makefile:23]: - [x] Raise an error if `NRFX_BASE` is not set.

File: src/HW_models/irq_ctrl.c
Comment (rugeGerritsen) [reply src/HW_models/irq_ctrl.c]: - [x] Remove extra empty line

**Comment by rugeGerritsen** [comment]
> Have you considered having all the original headers functions be compiled in but as weak (`__attribute__((weak))`), and then only providing a replacement implementation in the c files for the ones that need replacing?

This was a nice suggestion. I will give it a try:
1. Not defining `NRF_DECLARE_ONLY`
2. Defining `NRF_STATIC_INLINE` as `__attribute__((weak))`

File: CMakeLists.txt
Comment (rugeGerritsen) [reply CMakeLists.txt:25]: I did so by not adding that particular folder to the folders to be included. To safeguard this, I've now added a safeguard( `#ifdef __ZEPHYR__` ) in these.

**Comment by rugeGerritsen** [comment]
@aescolar . I've believe I've addressed your comments now. 

Using weak function definitions from the nrfx header files as a default did not really work out as the declaration and definition only use `NRF_STATIC_INLINE`.

File: Makefile
Comment (aescolar) [reply Makefile:23]: Hi @rugeGerritsen Please also describe it in the README.md

File: Makefile
Comment (aescolar) [inline Makefile]: Please don't create a new default rule. (The default rule is compile. This file should not add any rule anyhow)
Is there any reason to not use the type of check I proposed?

Comment (rugeGerritsen) [reply Makefile]: Without creating a rule, I got an error

**Comment by aescolar** [comment]
@rugeGerritsen  you seem to have forgotten this commit this time?

commit 92b73c8be4913894b0410c2d461cad8544c7dcfd
Author: Gerritsen, Rubin <rubin.gerritsen@nordicsemi.no>
Date:   Wed Sep 2 13:03:09 2020 +0200

    Add nrf_power.c. Only include the functionaly required to build a zephyr app

File: Makefile
Comment (aescolar) [reply Makefile]: Using what I wrote?

Comment (rugeGerritsen) [reply Makefile]: Arg. I indented the error message. That is why it went wrong

File: src/nrfx/hal/nrf_aar.c
Comment (aescolar) [inline src/nrfx/hal/nrf_aar.c]: } missing

**Comment by aescolar** [comment]
A minor thing, but which needs to be fixed: source files should be in the src/ folder, not in the top level. (meaning the 2 new headers)

**Comment by rugeGerritsen** [comment]
> A minor thing, but which needs to be fixed: source files should be in the src/ folder, not in the top level. (meaning the 2 new headers)

Does `src/nrfx_porting` sound ok to you?

**Comment by aescolar** [comment]
> > A minor thing, but which needs to be fixed: source files should be in the src/ folder, not in the top level. (meaning the 2 new headers)
> 
> Does `src/nrfx_porting` sound ok to you?

Maybe src/nrfx_replacements?

**Comment by aescolar** [comment]
> Using weak function definitions from the nrfx header files as a default did not really work out as the declaration and definition only use NRF_STATIC_INLINE.

Something like this seems to work: (these are 2 commits on top of your previous branches)
https://github.com/aescolar/ext_NRF52_hw_models/commit/696f9d23522ea8d56f394a104d580e0f2efda172
https://github.com/aescolar/hal_nordic/commit/485593fbb86ca0da87413528e98935f5dc1028b8

**Comment by rugeGerritsen** [comment]
> > Using weak function definitions from the nrfx header files as a default did not really work out as the declaration and definition only use NRF_STATIC_INLINE.
> 
> Something like this seems to work: (these are 2 commits on top of your previous branches)
> [aescolar@696f9d2](https://github.com/aescolar/ext_NRF52_hw_models/commit/696f9d23522ea8d56f394a104d580e0f2efda172)
> [aescolar/hal_nordic@485593f](https://github.com/aescolar/hal_nordic/commit/485593fbb86ca0da87413528e98935f5dc1028b8)

Thanks for providing the example. This reduced code duplication even more. On purpose I've not squashed the commits to make it easier to review. Let me know when you want me to squash them

File: docs/README.md
Comment (aescolar) [inline docs/README.md]: Would it be nicer to add before this sentence (and rephrase a bit):
"
When compiling this component using the provided Makefile (not with Zephyr's build system), the environment variable `NRFX_BASE` must be set to the path where a nrfx has been cloned. The nrfx must be at least version 2.3.0.
So for example, if the nrfx has been cloned as:
```bash
cd /some_path/nrfx/
git clone git@github.com:NordicSemiconductor/nrfx.git .
```
`NRFX_BASE` must be set as:
```bash
export NRFX_BASE=/some_path/nrfx/
```
"
(note this is markdown, so you need to look in the "source" of the comment)

File: src/nrfx/hal/nrf_hal_originals.c
Comment (aescolar) [inline src/nrfx/hal/nrf_hal_originals.c]: This line would not be applicable here, or?

File: src/nrfx/hal/nrf_hal_originals.c
Comment (aescolar) [inline src/nrfx/hal/nrf_hal_originals.c]: Maybe better a bit longer explanation like:
```
We want to use all real nRFx HAL definitions, while allowing some of them to be overridden.
We do this by including here all real nRFx HAL provided functions, but defining them as "weak".
This allows the HW models to provide their own version when needed due to HW models differences.
```

File: src/nrfx/hal/nrf_hal_originals.c
Comment (aescolar) [inline src/nrfx/hal/nrf_hal_originals.c]: Silly note: These 2 are not really needed (they dont harm either), as this is a .c file which is not going to be included by anybody else (I shouldn't have left them there, sorry)

File: src/nrfx_replacements/nrfx_config.h
Comment (aescolar) [inline src/nrfx_replacements/nrfx_config.h]: Maybe better
"This file should not be included in builds with Zephyr. Use the one provided by the nordic_hal zephyr module in that case"
?
(same in glue.h)

File: src/nrfx_replacements/nrfx_config.h
Comment (aescolar) [inline src/nrfx_replacements/nrfx_config.h]: I guess this comment is not really applicable anymore?

Anyhow, it would probably make sense to write here a bit  longer description of what we are doing. What do you think about something like:
```
When the real nRFx HAL headers are included, this header will be included from them, with the two definitions below.
* The normal users of the nRFx HAL will only get the prototype of the functions from the real headers.
* The HW models, will include again the real nRFx headers once more, but setting NRF_DONT_DECLARE_ONLY, and replacing NRF_STATIC_INLINE. In this way, the HW models will include 1 copy of the functions bodies, but as normal, globally available functions, instead of inlined versions. (See src/nrfx/hal/nrf_hal_originals.c for more info)
```
?

File: src/nrfx/hal/nrf_ccm.c
Comment (aescolar) [inline src/nrfx/hal/nrf_ccm.c]: Maybe don't move this one if not needed?

File: src/nrfx/hal/README.md
Comment (aescolar) [inline src/nrfx/hal/README.md]: There is a typo, but maybe better to expand a little bit like:
```
The HAL source files provided with these models implement
some modifications to the HAL interface []() needed due to the HW models peculiarities.
```

File: src/nrfx/hal/README.md
Comment (aescolar) [inline src/nrfx/hal/README.md]: Maybe better 
```
The original implementation of the HAL APIs from the real nrfx is included in `nrf_hal_originals.c`.
All these functions are labelled as weak, so they can be replaced as needed with new definition in `src/nrfx/hal/`
This removes the need for duplicating code.
```
?

File: src/HW_models/nrfx_bsim_redef.h
Comment (aescolar) [inline src/HW_models/nrfx_bsim_redef.h]: How come this set needs to be defined here? (naive question)

File: src/HW_models/nrfx_bsim_redef.h
Comment (aescolar) [inline src/HW_models/nrfx_bsim_redef.h]: NRF_SOC_IF_H => NRFX_BSIM_REDEF_H

File: src/HW_models/nrfx_bsim_redef.h
Comment (aescolar) [inline src/HW_models/nrfx_bsim_redef.h]: Maybe better:
"This file redefines macros from nrfx that need to be different for simulated devices." ?  (note the typo too)

File: src/HW_models/core_cm4.h
Comment (aescolar) [inline src/HW_models/core_cm4.h:34]: Same ifndef request as in the other PR

File: src/HW_models/nrfx_bsim_redef.h
Comment (aescolar) [inline src/HW_models/nrfx_bsim_redef.h]: #endif /* NRFX_BSIM_REDEF_H */

**Review by aescolar — COMMENTED** [review]
I think this is looking quite nicely. Just a bunch of minor things below.
But I would like also @anangl opinion.
From my point of view, the commits can be rebased.

File: src/HW_models/nrfx_bsim_redef.h
Comment (rugeGerritsen) [reply src/HW_models/nrfx_bsim_redef.h]: Some applications may access registers directly, i.e. `NRF_FICR->SOME_REGISTER` instead of using the nrfx hal functions. The redefinitions done here ensure that this statement points to a register in a mocked peripheral, and not to some invalid memory address.

File: docs/README.md
Comment (aescolar) [inline docs/README.md]: Please wrap lines in this file at ~80 chars
(same in the other README.md file)

File: src/nrfx/hal/README.md
Comment (aescolar) [inline src/nrfx/hal/README.md]: One extra space too much :)

File: docs/README.md
Comment (aescolar) [inline docs/README.md]: I missed this sentence when reviewing before (there is a "grammar typo" :) ) 
Anyhow, would it make sense to replace with
```
This repo contains both models of the NRF52 HW as well as some replacement nrfx
HAL functions. When used in combination with the real nrfx, these should enable code
meant for the nrfx to run without needing further changes.
This includes Zephyr SW.
```
?

**Comment by aescolar** [comment]
@anangl Are you happy with the changes in this PR and in https://github.com/zephyrproject-rtos/hal_nordic/pull/53 ?

**Review by aescolar — COMMENTED** [review]
Another request (hopefully the last): could you please add a warning in the commit message for Zephyr users, telling them this commit breaks backwards compatibility with Zephyr <= 2.4.0

**Comment by aescolar** [comment]
Thanks @rugeGerritsen . Apart from those 4 minor things these 3 PRs seems fine to me.
I would very much like @cvinayak & @anangl to comment or give their blessing.

With regard to merging, given the close dependency between the PRs, I would think it would be best to wait until the Zephyr merge window is open again, and then merge the nrfx changes and this one, and right after the Zephyr one (to avoid any Zephyr user trying to use this HW models version as they are not compatible with older releases).

**Comment by rugeGerritsen** [comment]
Thanks @aescolar for your valuable input for this PR. I believe we now have ended up with something that is more maintainable and that is easier to extend for future use cases. I agree that this should wait until the Zephyr merge window is open again.

In the last push I have addressed your last comments.

**Comment by aescolar** [comment]
Thanks to you @rugeGerritsen . Indeed this should save us some time in the future

File: CMakeLists.txt
Comment (anangl) [inline CMakeLists.txt]: This is redundant. In `nordic_hal` [we already have](https://github.com/zephyrproject-rtos/hal_nordic/blob/615c762a6d2e411d631cc6bd18fb45f892a5cbc3/CMakeLists.txt#L22):
```
zephyr_compile_definitions_ifdef(CONFIG_SOC_COMPATIBLE_NRF52832 NRF52832_XXAA)
```

File: src/HW_models/nrfx_bsim_redef.h
Comment (anangl) [inline src/HW_models/nrfx_bsim_redef.h]: This is a bit problematic. The `NRFX_IRQ_NUMBER_GET` macro (and consequently `NRFX_PERIPHERAL_ID_GET` as well) is intended to be used in cases when the `nrfx_get_irq_number` function cannot be called, .e.g. in some static structure initialization. Currently, there are no such uses of the macro in Zephyr, but we may hit this problem at some point. And I don't see an easy way of solving this at the moment.

File: src/nrfx/hal/nrf_hal_originals.c
Comment (anangl) [inline src/nrfx/hal/nrf_hal_originals.c]: I guess it was supposed to be  `MWU_PRESENT`. Or `NRF_MWU_BASE`?

File: src/nrfx_replacements/nrfx_config.h
Comment (anangl) [inline src/nrfx_replacements/nrfx_config.h]: Why not extract this part into a separate file in `HW_models/` and include it from here and from `nrfx_config.h` in the hal_nordic module? This could make maintenance a bit easier.

File: src/HW_models/nrfx_bsim_redef.h
Comment (rugeGerritsen) [reply src/HW_models/nrfx_bsim_redef.h]: @anangl , one option would be to NOT redefine the register base defines. Then we could keep the macro as is. However, then we need to make sure that no code is using these defines to access memory in some way

File: src/nrfx/hal/nrf_hal_originals.c
Comment (rugeGerritsen) [reply src/nrfx/hal/nrf_hal_originals.c]: - [x] Use `MWU_PRESENT`

File: src/nrfx_replacements/nrfx_config.h
Comment (rugeGerritsen) [reply src/nrfx_replacements/nrfx_config.h]: - [x] Extract `nrfx_config_bsim.h`

File: src/HW_models/nrfx_bsim_redef.h
Comment (aescolar) [reply src/HW_models/nrfx_bsim_redef.h]: >However, then we need to make sure that no code is using these defines to access memory in some way 

that is unlikely possible, as people may have their own app code that accesses some HW registers directly

Comment (rugeGerritsen) [reply src/HW_models/nrfx_bsim_redef.h]: @anangl , @aescolar . Do you believe we can keep it as suggested in the PR? This may give us some problems if this macro is used in the future

File: CMakeLists.txt
Comment (rugeGerritsen) [reply CMakeLists.txt]: Removed

**Comment by rugeGerritsen** [comment]
The last four commits can be squashed. Committed them separately to make it easier to review.

**Comment by aescolar** [comment]
@rugeGerritsen Shouldn't the nrf.h be in the nrf_replacements folder and that folder be in a `target_include_directories(zephyr_interface BEFORE INTERFACE` to ensure it is the first one included?
(The nrfx_bsim_redef.h, nrf_bsim_redef.h, nrf_config_bsim.h and nrfx_glue_bsim.h would also seem to fit better in the nrfx_replacements/ or nrfx/ folder? )

File: src/nrfx/hal/nrf_hal_originals.c
Comment (anangl) [inline src/nrfx/hal/nrf_hal_originals.c:17]: Wrap it in `#ifdef AAR_PRESENT`, for consistency.

File: src/HW_models/nrfx_bsim_redef.h
Comment (anangl) [inline src/HW_models/nrfx_bsim_redef.h]: This is not needed here. It is supposed to be included through `nrf.h`.

File: src/HW_models/nrfx_bsim_redef.h
Comment (anangl) [inline src/HW_models/nrfx_bsim_redef.h]: Those inclusions are done in `<drivers/nrfx_common.h>`. No need to repeat them here. Same applies to `nrf_bsim_redef.h`.

File: CMakeLists.txt
Comment (anangl) [inline CMakeLists.txt:24]: This should stay, but now contain only `src/HW_models/`.

**Comment by anangl** [comment]
> @rugeGerritsen Shouldn't the nrf.h be in the nrf_replacements folder and that folder be in a `target_include_directories(zephyr_interface BEFORE INTERFACE` to ensure it is the first one included?

@aescolar It cannot be `nrf_replacements/`, as it contains `nrfx_config.h` and `nrfx_glue.h`, which are not to be used with Zephyr.

**Comment by aescolar** [comment]
I'm ok with this (assuming the new commits are squashed where they fit).
@anangl are you happy with it?
(It would be  nice to finish this PR, it has been open for 1.5 months, and is not a small one)

File: src/HW_models/nrfx_glue_bsim.h
Comment (anangl) [inline src/HW_models/nrfx_glue_bsim.h]: The comment is unsuitable for `nrfx_bsim_redef.h`.

**Review by anangl — APPROVED** [review]

**Comment by aescolar** [comment]
@rugeGerritsen Could you then do that small change @anangl requested + squash as appropriate, & rebase the Zephyr PR https://github.com/zephyrproject-rtos/zephyr/pull/27882 ?

**Comment by rugeGerritsen** [comment]
> @rugeGerritsen Could you then do that small change @anangl requested + squash as appropriate, & rebase the Zephyr PR [zephyrproject-rtos/zephyr#27882](https://github.com/zephyrproject-rtos/zephyr/pull/27882) ?

Yes, I will, doing it now :)

File: src/HW_models/nrfx_glue_bsim.h
Comment (rugeGerritsen) [reply src/HW_models/nrfx_glue_bsim.h]: Fixed

**Review by aescolar — APPROVED** [review]

**Pull request merged**
